File: cli/src/__tests__/no-shell.test.ts
// @vitest-environment node
/**
 * Static canary: no shell-injection surface anywhere under cli/src/.
 *
 * Reviewer verifies by adding `exec('...')` in a scratch branch and
 * confirming this test fails. The architect doc §3.3 explains why
 * `shell: false` is a non-negotiable structural defense.
 */

import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { describe, expect, test } from 'vitest';

const FORBIDDEN: readonly { pattern: RegExp; name: string }[] = [
  // shell: true on a spawn/exec options object.
  { pattern: /\bshell\s*:\s*true\b/, name: 'shell: true' },
  // exec(...) / execSync(...) — even via child_process.exec.
  { pattern: /\bexec\s*\(/, name: 'exec(' },
  { pattern: /\bexecSync\s*\(/, name: 'execSync(' },
  // execFile/execFileSync are fine (argv-only), but we ban exec.
];

function* walkTs(dir: string): Generator<string> {
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const path = join(dir, entry.name);
    if (entry.isDirectory()) {
      if (entry.name === '__tests__') continue;
      yield* walkTs(path);
    } else if (
      entry.isFile() &&
      path.endsWith('.ts') &&
      !path.endsWith('.test.ts') &&
      !path.endsWith('.d.ts')
    ) {
      yield path;
    }
  }
}

describe('cli/src purity — no shell injection surface', () => {
  test('no `shell: true`, `exec(`, `execSync(` anywhere under cli/src/', () => {
    const root = join(process.cwd(), 'cli', 'src');
    const offenders: { file: string; api: string; line: number }[] = [];
    for (const file of walkTs(root)) {
      const lines = readFileSync(file, 'utf-8').split('\n');
      for (const { pattern, name } of FORBIDDEN) {
        for (let i = 0; i < lines.length; i++) {
          const line = lines[i] ?? '';
          const trimmed = line.trim();
          // Skip comments and JSDoc lines.
          if (trimmed.startsWith('//') || trimmed.startsWith('*')) continue;
          if (pattern.test(line)) {
            offenders.push({ file, api: name, line: i + 1 });
          }
        }
      }
    }
    if (offenders.length > 0) {
      const msg = offenders.map((o) => `  ${o.file}:${o.line}  uses ${o.api}`).join('\n');
      throw new Error(`Forbidden shell API in CLI:\n${msg}`);
    }
    expect(offenders).toEqual([]);
  });
});

File: src-tauri/src/no_shell_test.rs
// Static-analysis canary: no shell-injection surface anywhere under
// src-tauri/src/.
//
// Mirrors the Phase 10 cli/src/__tests__/no-shell.test.ts regression
// guard, restated for the Rust spawn side. The architect doc §4.3
// commits to argv-only `Command::new("node")` invocation; this test
// fails if a future contributor accidentally introduces a shell-out.
//
// Reviewer verifies by adding `Command::new("sh")` in a scratch
// branch and confirming this test fails.

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::{Path, PathBuf};

    fn forbidden() -> Vec<(&'static str, &'static str)> {
        vec![
            ("Command::new(\"sh\"", "Command::new(\"sh\""),
            ("Command::new(\"cmd\"", "Command::new(\"cmd\""),
            ("Command::new(\"cmd.exe\"", "Command::new(\"cmd.exe\""),
            ("Command::new(\"/bin/sh\"", "Command::new(\"/bin/sh\""),
            ("Command::new(\"bash\"", "Command::new(\"bash\""),
            // Powershell would also be a shell-out — guard.
            ("Command::new(\"powershell\"", "Command::new(\"powershell\""),
            ("Command::new(\"pwsh\"", "Command::new(\"pwsh\""),
        ]
    }

    fn collect_rs(dir: &Path, out: &mut Vec<PathBuf>) {
        let read = match fs::read_dir(dir) {
            Ok(r) => r,
            Err(_) => return,
        };
        for entry in read.flatten() {
            let path = entry.path();
            let name = entry.file_name();
            let name = name.to_string_lossy();
            if path.is_dir() {
                // Skip Cargo target and any hidden dirs.
                if name == "target" || name.starts_with('.') {
                    continue;
                }
                collect_rs(&path, out);
            } else if let Some(ext) = path.extension() {
                if ext == "rs" {
                    // Skip the canary itself.
                    if name == "no_shell_test.rs" {
                        continue;
                    }
                    out.push(path);
                }
            }
        }
    }

    #[test]
    fn rust_sources_have_no_shell_invocations() {
        let root: PathBuf = std::env::var("CARGO_MANIFEST_DIR")
            .map(PathBuf::from)
            .expect("CARGO_MANIFEST_DIR not set")
            .join("src");
        let mut files = Vec::new();
        collect_rs(&root, &mut files);
        assert!(
            !files.is_empty(),
            "expected to scan at least one .rs file under {}",
            root.display()
        );
        let mut offenders: Vec<String> = Vec::new();
        for path in &files {
            let text = match fs::read_to_string(path) {
                Ok(s) => s,
                Err(_) => continue,
            };
            for (line_no, line) in text.lines().enumerate() {
                let trimmed = line.trim_start();
                // Skip comments — JSDoc-equivalent and `//` notes can
                // legitimately reference the names without invoking
                // them.
                if trimmed.starts_with("//") {
                    continue;
                }
                for (name, needle) in forbidden() {
                    if line.contains(needle) {
                        offenders.push(format!(
                            "{}:{}  uses {}",
                            path.display(),
                            line_no + 1,
                            name
                        ));
                    }
                }
            }
        }
        assert!(
            offenders.is_empty(),
            "Forbidden shell invocation(s) in Rust source:\n{}",
            offenders.join("\n")
        );
    }
}
